File: Prova3/GodOfWar.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Scanner;
import java.util.Set;

public class GodOfWar {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int q = sc.nextInt();
        sc.nextLine(); // Consumir a linha restante

        Set<String> dictionary = new HashSet<>();
        for (int i = 0; i < q; i++) {
            dictionary.add(sc.nextLine());
        }

        int p = sc.nextInt();
        sc.nextLine(); // Consumir a linha restante

        List<String> words = new ArrayList<>();
        for (int i = 0; i < p; i++) {
            words.add(sc.nextLine());
        }

        String message = sc.nextLine();

        String[] messageWords = message.split(" ");
        StringBuilder correctedMessage = new StringBuilder();

        for (String word : messageWords) {
            if (dictionary.contains(word)) {
                correctedMessage.append(word).append(" ");
            } else {
                String bestMatch = null;
                int minDistance = Integer.MAX_VALUE;

                for (String dictWord : dictionary) {
                    int distance = levenshteinDistance(word, dictWord);
                    if (distance < minDistance || (distance == minDistance && dictWord.compareTo(bestMatch) < 0)) {
                        minDistance = distance;
                        bestMatch = dictWord;
                    }
                }

                correctedMessage.append(bestMatch).append(" ");
            }
        }

        System.out.println(correctedMessage.toString().trim());
        sc.close();
    }

    private static int levenshteinDistance(String a, String b) {
        int[][] dp = new int[a.length() + 1][b.length() + 1];

        for (int i = 0; i <= a.length(); i++) {
            for (int j = 0; j <= b.length(); j++) {
                if (i == 0) {
                    dp[i][j] = j;
                } else if (j == 0) {
                    dp[i][j] = i;
                } else {
                    dp[i][j] = Math.min(dp[i - 1][j - 1] 
                                 + (a.charAt(i - 1) == b.charAt(j - 1) ? 0 : 1), 
                                 Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1));
                }
            }
        }

        return dp[a.length()][b.length()];
    }
}

File: Prova3/GodOfWar.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_WORD_LENGTH 102

int min(int a, int b) {
    return (a < b) ? a : b;
}

int levenshteinDistance(char *a, char *b) {
    int len_a = strlen(a);
    int len_b = strlen(b);
    int dp[len_a + 1][len_b + 1];

    for (int i = 0; i <= len_a; i++) {
        for (int j = 0; j <= len_b; j++) {
            if (i == 0) {
                dp[i][j] = j;
            } else if (j == 0) {
                dp[i][j] = i;
            } else {
                dp[i][j] = min(dp[i - 1][j - 1] + (a[i - 1] != b[j - 1]), min(dp[i - 1][j] + 1, dp[i][j - 1] + 1));
            }
        }
    }

    return dp[len_a][len_b];
}

int main() {
    int q, p;
    scanf("%d", &q);
    char dictionary[q][MAX_WORD_LENGTH];
    for (int i = 0; i < q; i++) {
        scanf("%s", dictionary[i]);
    }

    scanf("%d", &p);
    char words[p][MAX_WORD_LENGTH];
    for (int i = 0; i < p; i++) {
        scanf("%s", words[i]);
    }

    getchar(); // Consumir a nova linha
    char message[10001];
    fgets(message, 10001, stdin);

    char *token = strtok(message, " ");
    char correctedMessage[10001] = "";

    while (token != NULL) {
        char *word = token;
        word[strcspn(word, "\n")] = '\0'; // Remover nova linha

        int found = 0;
        for (int i = 0; i < q; i++) {
            if (strcmp(word, dictionary[i]) == 0) {
                strcat(correctedMessage, dictionary[i]);
                strcat(correctedMessage, " ");
                found = 1;
                break;
            }
        }

        if (!found) {
            char *bestMatch = NULL;
            int minDistance = MAX_WORD_LENGTH;

            for (int i = 0; i < q; i++) {
                int distance = levenshteinDistance(word, dictionary[i]);
                if (distance < minDistance || (distance == minDistance && strcmp(dictionary[i], bestMatch) < 0)) {
                    minDistance = distance;
                    bestMatch = dictionary[i];
                }
            }

            strcat(correctedMessage, bestMatch);
            strcat(correctedMessage, " ");
        }

        token = strtok(NULL, " ");
    }

    correctedMessage[strlen(correctedMessage) - 1] = '\0'; // Remover espaço extra
    printf("%s\n", correctedMessage);

    return 0;
}
